### ISSUE TITLE
Conditional expression not working with defaults when met with other struct

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#436]
_Originally opened by @plaisted in https://github.com/cuelang/cue/issues/436_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
0.2.1
</pre>

### Does this issue reproduce with the latest release?
Yes

### What did you do?

Given
```
a: {
    val: *"default" | string
    result: *"not-matched" | string
    if (val == "match") {
        result: "matched"
    }
}

match: a & {
    val: "match"
}

not: a & {
    val: "other"
}
```

As expected this evals to:
```
a: {
    val:    "default"
    result: "not-matched"
}
match: {
    val:    "match"
    result: "matched"
}
not: {
    val:    "other"
    result: "not-matched"
}
```

However if I add a definition to `a` like:
```
#a: {
    val: string
    result: string
}

a: #a & {
    val: *"default" | string
    result: *"not-matched" | string
    if (val == "match") {
        result: "matched"
    }
}

match: a & {
    val: "match"
}

not: a & {
    val: "other"
}
```

It evals without ever matching the conditional for any of the cases:
```
#a: {
    val:    string
    result: string
}
a: {
    val:    "default"
    result: "not-matched"
}
match: {
    val:    "match"
    result: "not-matched"
}
not: {
    val:    "other"
    result: "not-matched"
}
```
Note this behavior stays to same if I use another struct instead of a definition for `#a` and regardless of where I define the defaults (on `#a` or `a`). It does work if defaults are not used. It also works if I simulate what I believe `#a &` is doing internally and set `a.val: string & (*"default" | string)`.

### What did you expect to see?
I expect the conditional to behave the same and match.result should eval to "matched" as it did without the definition.

### What did you see instead?
See above.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/436#issuecomment-654426232_

I verified that this is fixed with the new evaluator. So this can be expected to be fixed in v0.3.0 and earlier release candidates.

**Comment by cueckoo** [comment]
_Original reply by @plaisted in https://github.com/cuelang/cue/issues/436#issuecomment-656300531_

Sounds good. Is there a high level timeline regarding the new evaluators release or a branch I can build from source with it? The above functionality is core to a POC I was playing around with so I'll hold off on it until the evaluator is available.

**Comment by cueckoo** [comment]
_Original reply by @shykes in https://github.com/cuelang/cue/issues/436#issuecomment-666858206_

I just realized my issue #459 is a duplicate of this one.

FYI @plaisted there is now an alpha release of the new evaluator: https://github.com/cuelang/cue/releases/tag/v0.3.0-alpha1

I am going to check if it fixes the issue.

**Comment by cueckoo** [comment]
_Original reply by @shykes in https://github.com/cuelang/cue/issues/436#issuecomment-666860338_

Confirming that the problem is solved in 0.3.0-alpha1, at least for my repro case.

**Comment by cueckoo** [comment]
_Original reply by @plaisted in https://github.com/cuelang/cue/issues/436#issuecomment-670546406_

thanks for verifying, will go ahead an close this issue.
